New post: stable diffusion inference on Intel CPUs (#935)
huggingface/blog: stable_diffusion_inference_intel.md

Reviewer: i would use dashes not underscores

Reviewer: also maybe remove `inference`

Author: there's a fine-tuning post coming :)

---
title: "Accelerating Stable Diffusion Inference on Intel CPUs"
thumbnail: /blog/assets/xxx/01.png
authors:
- user: juliensimon
- user: ellacharlaix
---

> <h1>
> Accelerating Stable Diffusion Inference on Intel CPUs</h1>

Reviewer: ```suggestion
# Accelerating Stable Diffusion Inference on Intel CPUs
```

At the time of writing, the simplest way to get your hands on a Sapphire Rapids server is to use the Amazon EC2 [R7iz](https://aws.amazon.com/ec2/instance-types/r7iz/) instance family. As it's still in preview, you have to [sign up](https://pages.awscloud.com/R7iz-Preview.html) to get access. Like in previous posts, I'm using an `r7iz.metal-16xl` instance (64 vCPU, 512GB RAM) with an Ubunutu 20.04 AMI (`ami-07cd3e6c4915b2d18`).

Let's get started!

## The Diffusers library

The [Diffusers](https://huggingface.co/docs/diffusers/index) library makes it extremely simple to generate images with Stable Diffusion. If you're not familiar with these models, here's a great [illustrated introduction](https://jalammar.github.io/illustrated-stable-diffusion/).

First, let's create a virtual environment with the required libraries: Transformers, Diffusers, Accelerate, and PyTorch.

```
virtualenv sd_inference
source sd_inference/bin/activate
pip install transformers diffusers accelerate torch==1.13.1
```

Then, we write a simple benchmarking function that repeatedly runs inference, and returns the average latency for a single-image generation.

> ```

Reviewer: ```suggestion
```python
```

import time

def elapsed_time(pipeline, prompt, nb_pass=10, num_inference_steps=20):
	# warmup
	images = pipeline(prompt, num_inference_steps=10).images
	start = time.time()
	for _ in range(nb_pass):
		_ = pipeline(prompt, num_inference_steps=num_inference_steps, output_type="np")
	end = time.time()
	return (end - start) / nb_pass
```

Now, let's build a `StableDiffusionPipeline` with the default `float32` data type, and measure its inference latency.

```
import torch
from diffusers import StableDiffusionPipeline

prompt = "sailing ship in storm by Rembrandt"
model_id = "runwayml/stable-diffusion-v1-5"

pipe = StableDiffusionPipeline.from_pretrained(model_id).to("cpu")
latency = elapsed_time(pipe, prompt)
print(latency)
```

The average latency is **32.6 seconds**. As demonstrated by this [Intel Space](https://huggingface.co/spaces/Intel/Stable-Diffusion-Side-by-Side), the same code runs on a previous generation Intel Xeon (code name Ice Lake) in about 45 seconds. 

Out of the box, we can see that Sapphire Rapids CPUs are quite faster without any code change!

Now, let's accelerate!

## Optimum Intel and OpenVINO

> [Optimum Intel](https://huggingface.co/docs/optimum/intel/index) accelerates end-to-end pipelines on Intel architectures. Its API is extremely similar to the vanilla [Transformers](https://huggingface.co/docs/transformers/index) API, making it trivial to adapt existing code.

Reviewer 2: ```suggestion
[Optimum Intel](https://huggingface.co/docs/optimum/intel/index) accelerates end-to-end pipelines on Intel architectures. Its API is extremely similar to the [Diffusers](https://huggingface.co/docs/diffusers/index) API, making it trivial to adapt existing code.
```

Author: makes sense, done

Optimum Intel supports [OpenVINO](https://docs.openvino.ai/latest/index.html), an Intel open-source toolkit for high-performance inference. Let's update the code above: we only need to replace `StableDiffusionPipeline` with `OVStableDiffusionPipeline`:

> 

Reviewer 2: Optimum can be installed using pip as follows:

```
pip install optimum[openvino]
```

Author: done

```
from optimum.intel.openvino import OVStableDiffusionPipeline
...

> ov_pipe = OVStableDiffusionPipeline.from_pretrained(model_id, export=True).to("cpu")

Author: done

latency = elapsed_time(ov_pipe, prompt)
print(latency)

> ```

Author: done
